### ISSUE TITLE
Revisit Unit-testing framework

### ISSUE THREAD

**Issue by kdahlquist** [dondi/GRNsight#160]
Based on the discovery of unit test files in folder test-files > Data samples > module-tests, we should revisit the unit testing framework to see if those tests still work and then seeing what we can automate.

Once @dondi lines up some initial tests, he'll then show @NAnguiano and @anuvarsh how to do it.

Start a routine that before we commit code we will run the tests.

**Comment by dondi** [comment]
It's spring! We will use Mocha + Chai for our unit tests. The initial step is for @anuvarsh to start with a Mocha .js file containing a small number of tests. We don't currently know how to read files in the test, so @NAnguiano will break up the current code so that a function that takes the *parsed* file is accessible. @anuvarsh can then start writing unit tests with objects that match the parsed versions of equivalent Excel files.

Meanwhile, @dondi will investigate how the Excel files can be used directly within the framework.

**Comment by NAnguiano** [comment]
I broke up the spreadsheet controller so that the sheet is now passed into a function that takes in the parsed sheet instead of both actions being performed under the same function.

**Comment by anuvarsh** [comment]
I have created a demo version of a couple unit tests, but when I was trying to test their functionality with Mocha, I got an error saying that I didn't have permission to open the test.js file. Though this could be a number of things, I feel that I am most likely missing a naming convention that Mocha requires, and will continue to read into what could fix this.

**Comment by dondi** [comment]
In order to investigate this further, we'll need the code to be committed. Let's create a branch off beta called test-suite and place @anuvarsh's testing code/settings there. Once that branch is up and running we can merge back into beta.

**Comment by dondi** [comment]
To create a new branch in `git`:

    git checkout <originating branch>
    git checkout -b <branch name>

(work work work)

    git push origin <branch name>

The way to check if everything worked is to look in the branch dropdown on the website.

**Comment by anuvarsh** [comment]
The new test-suite branch is now on GitHub! I am going to go ahead and re-npm install mocha and chai in the correct folder to make sure they are reachable when running tests.

**Comment by anuvarsh** [comment]
Fun fact and reminder for the future: In order to "import" external files into mocha, you need to `require` the full path of the file that you want to import and save it as a variable. Then, every time you want to test a method within the imported file, you must first call the variable where the path is saved, and then call the method. 

An example edited from Stack Overflow (http://tinyurl.com/mocha-imports):

    var app = require('../app.js')
   
    it('should be true' function() {
       expect(app.foo()).to.be.true;
    });

**Comment by anuvarsh** [comment]
Mocha has been installed and saved in the test-suite branch. A test file has been uploaded as well, but currently there is just a demo test in there to make sure everything is working. The chai assert library has also been installed. 

Currently, the demo test that I have in the test.js file works when running mocha on command line:
![image](https://cloud.githubusercontent.com/assets/8509674/6094077/dc5461c8-aece-11e4-88f2-71baa0697f70.png)

However, we should be able to run tests by running `make test` so we can modify the reporter format and other things. This currently isn't working for me, but I am looking for solutions.

**Comment by dondi** [comment]
What error exactly are you getting for `make test`?

**Comment by anuvarsh** [comment]
![image](https://cloud.githubusercontent.com/assets/8509674/6094441/67eb278e-aedf-11e4-8ac3-6bd13ab75414.png)

But I'm using one of the examples that Mocha has online and this should be working.

**Comment by dondi** [comment]
Looks like `Makefile` is either out of sync or needs to be edited. Does the documentation imply that editing might be needed? If not, there might be a configuration issue. If so, then you may need to learn how to customize `Makefile`.

I looked at the file on GitHub and `test` relies on `test-unit`, which in turn calls some kind of `REPORTER` routine. Do any of those ring a bell? Maybe there is additional setup needed before that sequence can work.

**Comment by anuvarsh** [comment]
The Mocha website didn't mention the MakeFile, but after some research I found that this is something that other testers recommended to do. I found a couple different customizations to the MakeFile that other people have successfully been using. However, I do not know how to customize a MakeFile at the time, but can learn directly from @dondi or many other people in the CS department. 

I have a test written for a duplicate gene test sheet currently. Most online sources highly recommended having the functions that will be tested in the test file, so I copied the `parseSheet` function into the test file. I initially required the full path of the sheet, but I just found this to be more convenient because there are fewer variables to keep track of. 

The only hoop to jump through at this point is that the parsed sheet is a super long Object to include in the unit test. The array that holds the errors and the possible causes is much smaller, but I'm not sure if it would actually help to just input select parts of the parsed sheet into the unit test and then copy and paste the error message to check that they are equal. Another option would be to assign the parsed sheet to a variable and just use that variable in the unit test. It would make the meat of the unit test more readable, but there would still be huge Objects before a majority of the unit tests. Not sure if that's a problem.

**Comment by dondi** [comment]
Makefile issue solved; the need for learning `require` has been confirmed. We definitely do not want a copy of the code in the test; instead, figure out how `require` can be used to load the latest version of the code.

**Comment by anuvarsh** [comment]
Progress has been made in regards to how `require` is used based on some research of what big programs using Mocha (Express) have done, as well as what I found here: http://stackoverflow.com/questions/12258806/requiring-external-js-file-for-mocha-testing. A `module.export` for `parseSheet` was added because that is required to get the function to export. 

That being said, currently, the path isn't being read, and I am getting the error: `Cannot find module './server/controller/spreadsheet-controller.js'`.

**Comment by dondi** [comment]
We looked at it at the meeting and resolved some of the connect-the-dot issues between `require`, `module.export`, and `parseSheet`. Next immediate step is to actually invoke `parseSheet` from the test function with the desired sheet. A little more restructuring may be needed to separate `parseSheet` as an independently-callable function.

Reference for `requires` and `module.export`: http://nodejs.org/api/modules.html

**Comment by anuvarsh** [comment]
Currently, the obstacle to overcome is figuring out what exactly the Object `processGRNmap()` looks like since I made the mistake of putting the Object that is returned in the JavaScript console, but that is the Object returned by `parseSheet()`. Not completely sure where to find that Object, but still looking.

**Comment by anuvarsh** [comment]
Disregard that last comment, I misunderstood what was being returned in the JavaScript console, but was corrected. It is the `processGRNmap()` Object.

**Comment by anuvarsh** [comment]
Code has been updated to include some relevant parts of the sheet Object, but now the following error is being returned:
`Cannot find module: 'multiparty'`

**Comment by dondi** [comment]
Fortunately `Cannot find module` errors are generally solvable with an `npm install`. Going past that, we then needed to accurately replicate the code sequence of loading a file and feeding it into `parseSheet`. Finally, we then discovered that the latest version of `node-xlsx` appears to have changed its output format.

Follow-up tasks:
* Decide what to do with `node-xlsx` (we currently depend on `0.4.0` but the latest version is `0.5.1`).
* @anuvarsh can start implementing tests in full.
